src/termination.jl

```diff
@@ -0,0 +1,209 @@
+
+## General Arguments
+
+  * `abstol`: Absolute Tolerance
+  * `reltol`: Relative Tolerance
+
+## Arguments specific to `*_safe_*` modes
+
+  * `protective_threshold`: If the objective value increased by this factor wrt initial objective terminate immediately.
+  * `patience_steps`: If objective is within `patience_objective_multiplier` factor of the criteria and no improvement within `min_max_factor` has happened then terminate.
+
+"""
+struct SteadyStateTerminationCriteria{mode, T, S <:
+                                               Union{<:SafeTerminationOptions, Nothing}}
+    abstol::T
+    reltol::T
+
+    safe_termination_options::S
+end
+
+function Base.show(io::IO, s::SteadyStateTerminationCriteria{mode}) where {mode}
+    print(io,
+          "SteadyStateTerminationCriteria(mode = $(mode), abstol = $(s.abstol), reltol = $(s.reltol)")
+    if mode ∈ (:rel_safe, :rel_safe_best, :abs_safe, :abs_safe_best)
+        print(io, ", safe_termination_options = ", s.safe_termination_options, ")")
+    else
+        print(io, ")")
+    end
+end
+
+function SteadyStateTerminationCriteria(mode = :default; abstol::T = 1e-8,
+                                        reltol::T = 1e-6, protective_threshold = 1e3,
+                                        patience_steps::Int = 30,
+                                        patience_objective_multiplier = 3,
+                                        min_max_factor = 1.3) where {T}
+    safe_termination_options = if mode ∈
+                                  (:rel_safe, :rel_safe_best, :abs_safe, :abs_safe_best)
+        SafeTerminationOptions(protective_threshold, patience_steps,
+                               patience_objective_multiplier, min_max_factor)
+    else
+        nothing
+    end
+    return SteadyStateTerminationCriteria{mode, T, typeof(safe_termination_options)}(abstol,
+                                                                                     reltol,
+                                                                                     safe_termination_options)
+end
+
+_get_termination_mode(::SteadyStateTerminationCriteria{mode}) where {mode} = Val(mode)
+
+for mode in (:default, :norm, :rel, :rel_norm, :abs, :abs_norm)
+    T = SteadyStateTerminationCriteria{mode}
+    mode_val = Val(mode)
+    @eval function _get_termination_condition(::$(T), storage = nothing)
+        function _termination_condition_closure(integrator, abstol, reltol, min_t)
+            return _has_converged(get_du(integrator), integrator.u, $(mode_val), abstol,
+                                  reltol)
+        end
+    end
+end
+
+for mode in (:rel_safe, :rel_safe_best, :abs_safe, :abs_safe_best)
+    T = SteadyStateTerminationCriteria{mode}
+    mode_val = Val(mode)
+    @eval function _get_termination_condition(cond::$(T), storage)
```

Comment:
do they need to be dispatches?

Reply:
It feels wrong to have something that is known at compile time and make them runtime branches :sweat_smile:

Reply:
but is this improving runtime or just adding compile time?

Reply:
This particular set of `@eval`s isn't adding any compile time. But the `_has_converged` ones add like `0.8ms` over `3.6ms` compilation time.

Reply:
I still think it's a hyperoptimization to make this all dispatch, but we can keep it for now.